### ISSUE TITLE
`isinstance(treespec, LeafSpec)` is deprecated, use `isinstance(treespec, TreeSpec) and treespec.is_leaf()` instead

### ISSUE THREAD

**Issue by mscheltienne** [Lightning-AI/pytorch-lightning#21326]
### Bug description

The following `FutureWarning` appeared in our CIs:

```
tests/integration/lightning/test_training_integration.py:415: in test_training_with_different_optimizers
    trainer.fit(model, train_loader, val_loader)
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/trainer.py:560: in fit
    call._call_and_handle_interrupt(
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/call.py:49: in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/trainer.py:598: in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/trainer.py:1011: in _run
    results = self._run_stage()
              ^^^^^^^^^^^^^^^^^
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/trainer.py:1053: in _run_stage
    self._run_sanity_check()
.venv/lib/python3.13/site-packages/lightning/pytorch/trainer/trainer.py:1082: in _run_sanity_check
    val_loop.run()
.venv/lib/python3.13/site-packages/lightning/pytorch/loops/utilities.py:179: in _decorator
    return loop_run(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.13/site-packages/lightning/pytorch/loops/evaluation_loop.py:120: in run
    self.setup_data()
.venv/lib/python3.13/site-packages/lightning/pytorch/loops/evaluation_loop.py:180: in setup_data
    combined_loader = CombinedLoader(dataloaders, "sequential")
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.13/site-packages/lightning/pytorch/utilities/combined_loader.py:288: in __init__
    self._flattened, self._spec = _tree_flatten(iterables)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.13/site-packages/lightning/pytorch/utilities/_pytree.py:21: in _tree_flatten
    return [pytree], LeafSpec()
                     ^^^^^^^^^^
../../_temp/uv-python-dir/cpython-3.13.9-linux-x86_64-gnu/lib/python3.13/warnings.py:594: in __new__
    warn(msg, category=category, stacklevel=stacklevel + 1)
E   FutureWarning: `isinstance(treespec, LeafSpec)` is deprecated, use `isinstance(treespec, TreeSpec) and treespec.is_leaf()` instead.
```

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

N/A

### How to reproduce the bug

```python
N/A
```

### Error messages and logs

N/A

### Environment

<details>
  <summary>Current environment</summary>

```
lightning:              2.5.5
torch:                  2.10.0.dev20251103+cpu
```

</details>

### More info

N/A

cc @ethanwharris

**Comment by SkafteNicki** [comment]
Hi @mscheltienne, thanks for raising this issue. I am trying to reproduce the issue and is not able to do. However, I can see you are running torch 2.10 (dev branch) so this is probably the reason, I will investigate further.

**Comment by mscheltienne** [comment]
Yes, this warning appeared in our `pre-release` CI which tests our dependencies against their nightly release. I suspect `torch` changed something that is now yielding this warning.

**Comment by intexcor** [comment]
same

**Comment by intexcor** [comment]
torch 2.10 released

**Comment by etdufay** [comment]
Same

**Comment by bhimrazy** [comment]
Hi all, thanks for the reports! 🙏

While we work on the fix and prepare a new release, I think it would be good to pin torch to  `< 2.10`.
Alternatively, if ignoring this specific FutureWarning works for you in the meantime, that’s fine as well.

**Comment by deependujha** [comment]
the warning is only for asserting instance type, and no breaking changes. It's completely safe to ignore the warning, as lightning nowhere uses `isinstance(treespec, LeafSpec)`.

PR #21502 is expected to be landed soon.

**Comment by adamjstewart** [comment]
#21502 fixes Lightning's CI but it doesn't fix the hundreds of other libraries that depend on Lightning.

**Comment by mscheltienne** [comment]
#21502 Ignores the deprecation? It doesn't fix the usage of a deprecated element here:

```
.venv/lib/python3.13/site-packages/lightning/pytorch/utilities/_pytree.py:21: in _tree_flatten
    return [pytree], LeafSpec()
```

And hides it, so now if we forget about it when it gets removed, `lightning` will be broken?

**Comment by deependujha** [comment]
hi @mscheltienne

`LeafSpec` itself is not deprecated. The PyTorch deprecation only applies to `isinstance(treespec, LeafSpec)` checks, not constructing or returning `LeafSpec()` objects.

In Lightning, we don’t rely on `isinstance(treespec, LeafSpec)` anywhere; the warning is triggered internally during doctests via PyTorch’s decorator. That’s why filtering the warning unblocks CI without masking deprecated Lightning usage.

That said, I agree we should not lose track of this. I’m happy to add a short comment / TODO referencing the PyTorch deprecation so we can revisit once the API is fully stabilized upstream.

https://github.com/pytorch/pytorch/blob/ed18b31a89b1764bea59cef6786747f376992fde/torch/utils/_pytree.py#L1331-L1350

**Comment by mscheltienne** [comment]
Oh apologies. I missed that entirely, it's quite strange to emit a deprecation warning on an operation which is not deprecated :/

**Comment by deependujha** [comment]
Hi @adamjstewart,

That’s correct, this change is intentionally scoped to unblocking Lightning’s CI.

The warning originates from PyTorch internals and is emitted when `LeafSpec()` is constructed. Lightning itself does not rely on deprecated semantics, but our doctests trigger the warning, which causes CI failures under strict warning settings.

Downstream libraries depending on Lightning should not see additional warnings unless they independently construct `LeafSpec()` or run similar doctests against PyTorch internals. Addressing the warning at the ecosystem level would require an upstream PyTorch change, which is outside the scope of this PR.
